core/deployment/src/main/java/io/quarkus/deployment/steps/ReflectiveHierarchyStep.java

```diff
@@ -249,11 +259,30 @@ private void addClassTypeHierarchy(CombinedIndexBuildItem combinedIndexBuildItem
                     method.returnType().kind() == Kind.VOID) {
                 continue;
             }
-            visits.addLast(() -> addReflectiveHierarchy(combinedIndexBuildItem, reflectiveHierarchyBuildItem, source,
+            visits.addLast(() -> addReflectiveHierarchy(combinedIndexBuildItem, capabilities,
+                    reflectiveHierarchyBuildItem, source,
                     method.returnType(),
                     processedReflectiveHierarchies,
                     unindexedClasses, finalFieldsWritable, reflectiveClass, visits));
         }
+
+        // for Kotlin classes, we need to register the nested classes as well because companion classes are very often necessary at runtime
+        if (capabilities.isPresent(Capability.KOTLIN) && isKotlinClass(info)) {
+            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            try {
+                Class<?>[] declaredClasses = classLoader.loadClass(info.name().toString()).getDeclaredClasses();
+                for (Class<?> clazz : declaredClasses) {
+                    DotName dotName = DotName.createSimple(clazz.getName());
+                    addClassTypeHierarchy(combinedIndexBuildItem, capabilities, reflectiveHierarchyBuildItem, source,
+                            dotName, dotName,
+                            processedReflectiveHierarchies, unindexedClasses,
+                            finalFieldsWritable, reflectiveClass, visits);
+                }
+            } catch (ClassNotFoundException e) {
+                log.warnf(e, "Failed to load Class %s", info.name().toString());
+            }
+
+        }
```

Comment:
Can we move it to the `kotlin` extension? It feels weird to have this in core

Reply:
Not really, the handling of reflection is in the core

Reply:
Perhaps the `kotlin` extension may produce additional build items for this step to consume?

Reply:
The way I see it, that's really too much work for almost no gain. 
I'm certainly not going to do it, but I won't stop anyone else from taking it up :)